### PR TITLE
Audio system and asset rework

### PR DESCRIPTION
**Pull request by sornas**
Well, this took too long and got too big. The audio part is far from finished
but at this point I'd rather get the infrastructure merged.

## Audio

- Load f32 48000Hz stereo WAV.
- Play multiple sounds and the same sound concurrently.
- Reload sound assets.

## Assets

- Add asset IDs
- Get loaded assets via indexing. (See main.rs)

### CODE REVIEW

File: src/asset/mod.rs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+use std::time::SystemTime;
+
+/// A marker type for the unit pixels.
+pub type Pixels = usize;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ImageAssetID(usize);
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct AudioAssetID(usize);
```
Comment (sornas) [inline src/asset/mod.rs:18]: Newtype here ensures different assets aren't mixed.

File: src/asset/mod.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    fn try_get(&self, id: ImageAssetID) -> Option<&Image> {
+        self.images.get(id.0)
+    }
+}
+
+impl TryGetAsset<Audio, AudioAssetID> for AssetSystem {
+    fn try_get(&self, id: AudioAssetID) -> Option<&Audio> {
+        self.audio.get(id.0)
+    }
+}
```
Comment (sornas) [inline src/asset/mod.rs:92]: I'm split about including this. Due to reasons far above my type system knowledge, we can't have the index operator above return a `Option<&T>`, so this is a way of getting an asset and not panicking. But that only makes sense with dynamic loading of assets. So, keep or no? I guess it doesn't hurt anyone and we can still index to get assets unconditionally.

**Comment by sornas** [comment]
If the API looks Ok I can add some documentation before merging.

File: src/audio.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    position: usize,
+    looping: bool,
+    samples: Samples,
+
+    /// Remove this source when we get the opportunity.
+    ///
+    /// Gets set if
+    /// a) the audio is done playing and it doesn't loop,
+    /// b) it is requested by the user.
+    remove: bool,
```
Comment (FredTheDino) [inline src/audio.rs:18]: Maybe combine looping and remove into an enum?

File: src/asset/mod.rs (hunk 2 above)
Comment (FredTheDino) [reply src/asset/mod.rs:98]: Decided to remove.

**Review by FredTheDino — APPROVED** [review]

**Pull request merged**